src/phongo_bson.c

```diff
@@ -1035,9 +1126,44 @@ bool php_phongo_bson_to_zval_ex(const unsigned char* data, int data_len, php_pho
 			break;
 
 		case PHONGO_TYPEMAP_CLASS: {
-			zval obj;
+			zval              obj;
+			zend_class_entry* obj_ce = state->odm ? state->odm : state->map.root;
+
+#if PHP_VERSION_ID >= 80100
+			/* Enums require special handling for instantiation */
+			if (obj_ce->ce_flags & ZEND_ACC_ENUM) {
+				int       plen;
+				zend_bool pfree;
+				char*     case_name;
+				zval*     enum_case;
+
+				case_name = php_array_fetchc_string(&state->zchild, "name", &plen, &pfree);
+
+				if (!case_name) {
+					phongo_throw_exception(PHONGO_ERROR_UNEXPECTED_VALUE, "Missing 'name' field to infer enum case for %s", ZSTR_VAL(obj_ce->name));
+
+					goto cleanup;
+				}
+
+				enum_case = resolve_enum_case(obj_ce, case_name);
+
+				if (pfree) {
+					efree(case_name);
+				}
+
+				if (!enum_case) {
+					/* Exception already thrown */
+					goto cleanup;
+				}
+
+				ZVAL_COPY_VALUE(&obj, enum_case);
```

Comment:
I'm wondering what you expect to happen in the `bsonUnserialize` call for a typical unserializable enum case.  Since these are essentially singleton instances any side effects will be essentially global.

Reply:
I mentioned this in one of the test file comments, but there's really no practical purpose for `bsonUnserialize`. Since enums have no state, there's nothing that would be initialized. We're just invoking it because of the existing contract. The example implementation I used in the test file was to log a message that `bsonUnserialize` was called.

I think it sucks that we need to require enums to implement an interface, but I'd be more concerned about implicitly encoding their class name (i.e. pretending they're Persistable w/o the explicit implementation). The trait that @alcaeus suggested is probably the best compromise.

Reply:
FYI: a PersistableEnum trait with a no-op `bsonUnserialize()` method has been created.